### DISCUSSION TITLE
Type checking runtime config inside next.config.js

### DISCUSSION THREAD

**Discussion by bring-shrubbery** [vercel/next.js#47324] [Help]
## What I've got working

I want my calls to `getConfig` to be type checked. My current approach is to define a `config.d.ts` file and declare following in there:

```ts
declare module 'next/config' {
  export interface PublicRuntimeConfig {
    runtimeProperty: string
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  export interface ServerRuntimeConfig {
    serverProperty: number;
  }

  export default function getConfig(): {
    publicRuntimeConfig: PublicRuntimeConfig
    serverRuntimeConfig: ServerRuntimeConfig
  }
}

```

This makes any call to `getConfig` typed, for example this shows error as expected:

```ts
import getConfig from "next/config"

const { publicRuntimeConfig } = getConfig()

publicRuntimeConfig.asdf
// Property 'asdf' does not exist on type 'PublicRuntimeConfig'.
```

## Where I need help

Now what I'm trying to do, is to use this type definition in my `next.config.js` file using `// @ts-check` comment for type checking. I know it's not the same as full on TypeScript file.

I've tried typing it using :

```js
// @ts-check

/** @type {import("next").NextConfig} */
const config = {
  images: {
    domains: ["google.com"]
  },
  /** @type {import("next/config").PublicRuntimeConfig} */
  //                                   ^ Namespace '"./node_modules/next/config"' has no exported member 'PublicRuntimeConfig'
  publicRuntimeConfig: {
    ...
  }
}

module.exports = config;
```

The main confusion point for me is that in my code, that type import (`import type { PublicRuntimeConfig } from "next/config"`) works fine, but in `next.config.js` it doesn't recognise the defined types.

Is there something I'm doing wrong or is this expected behaviour?

**Comment by icyJoseph** [comment]
I think here the issue is that `/** @type {import("next").NextConfig} */`, defines its own interface for the public runtime config, and that `import("next/config")`, is just a function that exposes a singleton, filled in by the framework, with the values from next.config.js...

There's no coupling between them.

- https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/runtime-config.ts

So you kind of want to extend `/** @type {import("next").NextConfig} */` interface, so that it uses your public runtime config. That is you need a second extension here, but I don't remember how to extend this config here... 😓 😅

**Reply by bring-shrubbery** [reply]
Yeah, that's what I was trying to explain, maybe I did a bad job 😃 

My idea was to add the `/** @type {import("next/config").PublicRuntimeConfig} */` to the `next.config.js` file's `publicRuntimeConfig` field, so that it's being type-checked from the `config.d.ts` file that **I defined**. Unfortunately for some reason `// @ts-check` doesn't pick up the `d.ts` files that extend installed npm modules.

The `next.config.js` would ideally do nothing in definition of types, but instead would use the type that I defined outside of it, to make sure that both files are in sync.
